csrc/scheduler/normalization_inner.cpp

```diff
@@ -252,6 +261,386 @@ int64_t getMaxPersistentBatch(
+  // register_overhead is all registers except those for the persistent
+  // buffers. The register in each thread = register_overhead +
+  // persistent_buffer_size / bytes_per_register
+  // Current values are based on tests of sofmax, layer_norm, softmax_dropout,
+  // dropout_layer_norm on A100 & H100. It directly affects maxregcount passed
+  // to NVRTC and influences the occupancy.
+  const int64_t register_overhead = has_exp_op ? 32l : 16l;
+
+  // Target occupancy required to hide memory latency.
+  // Used to calculate the maximum register count each thread can use.
+  // Used to calculate the maximum persistent batch size.
+  // Current value is based on tests of sofmax, layer_norm, softmax_dropout,
+  // dropout_layer_norm on A100 & H100.
+  const int64_t target_warps_per_sm = 28l;
+
+  // device properties
+  const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  const int64_t threads_per_warp = (int64_t)dev_prop->warpSize;
+  const int64_t max_threads_in_block = (int64_t)dev_prop->maxThreadsPerBlock;
+  const int64_t max_threads_per_sm =
+      (int64_t)dev_prop->maxThreadsPerMultiProcessor;
+  const int64_t device_multiprocessor_count =
+      (int64_t)dev_prop->multiProcessorCount;
+
+  // alwasy use [vectorize_factor]
+  NVF_CHECK(
+      inner_most_dimension_numel % vectorize_factor == 0,
+      "Element count should be divisible by vectorize factor! count= ",
+      inner_most_dimension_numel,
+      ", vectorize_factor= ",
+      vectorize_factor);
+  const int64_t parallel_after_vectorize =
+      inner_most_dimension_numel / (int64_t)vectorize_factor;
+
+  // try to use at least 4 warps per block
+  const int64_t min_threads_per_block = 4l * threads_per_warp;
+
+  // set the min persistent buffer size to avoid requesting
+  // a block size larger than device limit
+  const int64_t batches_per_block_inner_reduction_min =
+      ceilDiv(parallel_after_vectorize, max_threads_in_block);
+
+  // set the max persistent batch size to avoid low occupancy
+  // don't increase persistent batch size until we have enough threads
+  const int64_t pbs_max_1 =
+      ceilDiv(parallel_after_vectorize, min_threads_per_block);
+  // derived the maximum persistent batch size from the target occupancy
+  const int64_t buffer_bytes_per_batch = max_persistent_buffer_size /
+      total_reduction_numel * (int64_t)vectorize_factor;
+  const int64_t target_threads_per_sm =
+      std::min(target_warps_per_sm * threads_per_warp, max_threads_per_sm);
+  const int64_t pbs_max_2 = getMaxPersistentBatch(
+      buffer_bytes_per_batch, target_threads_per_sm, register_overhead);
+  const int64_t batches_per_block_inner_reduction_max = std::max(
+      batches_per_block_inner_reduction_min, std::min(pbs_max_1, pbs_max_2));
+
+  // Compute maximum number of reductions we could do in the same kernel based
+  // on persistent buffer size. Bounded by the wave count for utilization of
+  // SMs.
+  const int64_t max_multi_reduction_factor = std::min(
```

Comment:
nitpick: not a super big fan of duplicating some of the logic here with the old code path.

Including: `register_overhead`, `targets_warps_per_sm`, computation of `buffer_bytes_per_batch`, `getMaxPersistentBatch`, `max_multi_reduction_factor`.

Do you think it make sense to unify those?

Reply:
Make sense.
The plan is to separate 2D inner persistent `(e.g. layerNorm, [Iter,Redu])` and 3D inner persistent `(e.g. batchNorm, [Redu, Iter, Redu])`, so these duplicated variables may change in the 3D inner persistent heuristic (not revised yet).

The solution I have is we can refactor this `namespace` into a class, and these vars are converted to priviate members so we don't need to duplicate or passing them around different functions. I'll make these changes in a single commit, if you feel the change is too big, we can drop it and redo in another PR.

Reply:
made a [draft](https://github.com/NVIDIA/Fuser/pull/1873/files/452d5d9dc65556c4ab1823a3e43eaa7475c3db20..7d032b66cba034fdddec655efe96eb21d9f97f09) to solve this issue: